src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs

```diff
@@ -0,0 +1,135 @@
+            {
+                throw new ArgumentNullException(nameof(solutionManager));
+            }
+
+            _settings = new Lazy<ISettings>(() => settings);
+            _solutionManager = new Lazy<IVsSolutionManager>(() => solutionManager);
+            _getLockFileOrNullAsync = getLockFileOrNullAsync ?? BuildIntegratedProjectUtility.GetLockFileOrNull;
+        }
+
+        public async Task<IVsPathContext> CreateAsync(Project project, CancellationToken token)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            // Try to read the path context from the assets file.
+            var outputPathContext = await GetPathContextFromAssetsFileAsync(project);
+
+            // Fallback to reading the path context from the solution's settings. Note that project level settings in
+            // VS are not currently supported.
+            if (outputPathContext == null)
+            {
+                var internalPathContext = NuGetPathContext.Create(_settings.Value);
+
+                outputPathContext = new VsPathContext(
+                    internalPathContext.UserPackageFolder,
+                    internalPathContext.FallbackPackageFolders);
+            }
+
+            return outputPathContext;
+        }
+
+        private async Task<IVsPathContext> GetPathContextFromAssetsFileAsync(Project project)
+        {
+            var nuGetProject = await _solutionManager.Value.GetOrCreateProjectAsync(
+                project,
+                new VSAPIProjectContext());
+
+            // It's possible the project isn't a NuGet-compatible project at all.
+            if (nuGetProject == null)
+            {
+                return null;
+            }
+
+            // It's possible that this project isn't a build integrated NuGet project at all. That is, this project may
+            // be a packages.config project.
+            var buildIntegratedProject = nuGetProject as BuildIntegratedNuGetProject;
+            if (buildIntegratedProject == null)
+            {
+                return null;
+            }
+
+            // It's possible the lock file doesn't exist or it's an older format that doesn't have the package folders
+            // property persisted.
+            var lockFile = await _getLockFileOrNullAsync(buildIntegratedProject);
+            if (lockFile == null ||
+                lockFile.PackageFolders == null ||
+                lockFile.PackageFolders.Count == 0)
+            {
+                return null;
+            }
+
+            // The user packages folder is always the first package folder. Subsequent package folders are always
+            // fallback package folders.
+            var packageFolders = lockFile
+                .PackageFolders
+                .Select(lockFileItem => lockFileItem.Path)
+                .ToList();
+
+            var userPackageFolder = packageFolders[0];
```

Comment:
There is a slightly possibility that someone restored with an older nuget.exe that wasn't writing this property out, which would be an out of bounds exception here